### DISCUSSION TITLE
[PROPOSAL] Fastlane Swift Improvements

### DISCUSSION THREAD

**Discussion by mollyIV** [fastlane/fastlane#22160] [Ideas]
## Introduction 

This proposal introduces enhancements to Fastlane in Swift, aiming to improve the development experience, simplify implementation, and lower the entry barrier for new contributors.

## Motivation 

Fastlane in Swift allows developers to write lanes using the Swift language, offering significant benefits such as faster development by catching errors at the compilation stage and improved collaboration. Since Apple platform developers are generally more familiar with Swift, using it for Fastlane can lead to increased contributions from the team. Additionally, it provides opportunities to leverage existing Swift packages for custom workflows.

However, the current Fastlane in Swift development experience is not ideal. Setting it up and running requires effort, whether using Xcode or Swift Package Manager (SPM). The former is now considered obsolete, and the unopinionated approach creates friction.

Since the feature was implemented long ago, many internal libraries used are now outdated or abandoned, adding unnecessary complexity and discouraging new contributors. Moreover, the Swift language has evolved significantly, providing better tools to enhance this feature, which will not only improve overall performance but also lower the barrier for new contributors.

Lastly, Fastlane in Swift sometimes does not respect the Fastlane version set by environment managers like `mise` in certain cases.

## Proposed solution 

The proposed solution focuses on improving the following aspects of the development experience: Setup, Core Implementation, and Public APIs. This proposal details the Setup design, while other aspects will be briefly mentioned to present the bigger picture. Follow-up proposals will address these aspects in detail.

> Note: The proposed solution may introduce breaking changes, but this should not be a problem since Fastlane in Swift is still in beta.

### Setup

Setting up Fastlane in Swift will be simplified, and Xcode project support will be discontinued. After executing the `fastlane init swift` command, a new Swift package will be created and ready for use.

The newly created Swift package will have the following structure:

```swift
// swift-tools-version: 5.9
import PackageDescription

let package = Package(
  name: "fastlane-swift",
  platforms: [.macOS(.v10_15)],
  dependencies: [
    .package(url: "https://github.com/apple/swift-argument-parser", from: "1.5.0"),
    .package(name: "Fastlane", url: "https://github.com/fastlane/fastlane", from: "2.x.y"),
  ],
  targets: [
    .executableTarget(
      name: "fastlane-swift",
      dependencies: [
        .product(name: "ArgumentParser", package: "swift-argument-parser"),
        .product(name: "Fastlane", package: "fastlane"),
      ]
    )
  ]
)
```

The `swift-argument-parser` will be used to parse the command-line arguments. 

The package's entry point will be updated as follows:

```swift
import ArgumentParser

@main
struct Fastlane: ParsableCommand {
  static let configuration = CommandConfiguration(
    abstract: "Peforming fastlane operations.",
    subcommands: [Lane.self]
  )
}

extension Fastlane {
  struct Lane: ParsableCommand {
    static let configuration = CommandConfiguration(abstract: "Runs the given lane.")
        
    @Argument(help: "The name of a lane to execute.")
    var name: String
        
    @Argument(help: "The parameters.")
    var params: [String] = []
        
    mutating func run() throws {
        Main().run(with: Fastfile())
    }
}
```

The CLI interface will remain the same:

```bash
$ swift run fastlane-swift lane foobar
``` 

### Core Implementation

The current implementation of Fastlane in Swift is fairly complex. It uses the [SwiftShell](https://github.com/kareman/SwiftShell) package, which was last updated four years ago and appears to be abandoned. Fastlane in Swift uses sockets internally to run lanes in parallel, but we could leverage Swift concurrency and allow the operating system to manage that.

The Swift language has evolved significantly, and the Foundation framework now provides excellent APIs to spawn and manage processes. Fastlane in Swift can greatly benefit from these advancements.

To get a glimpse of how simple running system processes can be these days, please refer to the [Command](https://github.com/tuist/Command) Swift package.

### Public APIs

Fastlane in Swift APIs are auto-generated based on existing Ruby APIs. For example:

```swift
println(message: "Hello, world!")

// Fastlane.swift 
public func println(message: OptionalConfigValue<String?> = .fastlaneDefault(nil)) { ... }
```

Fastlane APIs may seem overwhelming, and it’s not always easy to discover a lane to do the job. There are [ongoing discussions](https://github.com/fastlane/fastlane/discussions/20463#discussioncomment-3139084) on how the APIs can be improved in Fastlane 3.0.

Fastlane in Swift opens up unique opportunities to improve APIs by using a DSL.

We could start by grouping functionalities, such as building, testing, and uploading. Each group would be provided by a separate target.

```swift
import Building 

let build = Build(scheme: "my-scheme", device: .iPhone15, os: .v15)
try await BuildRunner(build: build).run()
```

Building (aka. scan) seems to be a great candidate to start with.

A DSL would improve the development experience, lane discoverability, and eliminate invalid states by restricting the parameters passed into the constructor/method. This would also help to test how Fastlane 3.0 could look.

## Future directions

**Lane Discoverability** 

Currently, lanes must follow a naming convention, i.e., `foobarLane()`. We could replace this naming convention with an explicit Swift macro:

```swift
@Lane
func foobar() { ... }
```

**DLS**
More Fastlane tools would be migrated to the DSL.

## Summary 

This proposal outlines significant enhancements to Fastlane in Swift, focusing on improving the development experience, simplifying implementation, and lowering the entry barrier for new contributors. Key improvements include:

* **Setup Simplification:** Streamlining the setup process by discontinuing Xcode project support and utilizing Swift Package Manager for easier initialization and use.
* **Core Implementation:** Updating the core implementation to leverage modern Swift concurrency features and Foundation APIs for process management, replacing outdated dependencies like SwiftShell.
* **Public APIs:** Introducing a more intuitive and discoverable DSL for Fastlane in Swift, allowing for improved grouping of functionalities and enhanced developer experience. The DSL will be iteratively improved based on user feedback.

Future directions include improving lane discoverability with Swift macros and expanding the DSL to encompass more Fastlane tools. These changes aim to make Fastlane in Swift more accessible, efficient, and user-friendly, setting the stage for a robust Fastlane 3.0.

Thanks for reading!

**Comment by pepicrft** [comment]
It's amazing seeing the community moving the needle to meet developers where they are allowing them to express their automation needs in Swift.

When I think about Fastlane's strengths in Ruby, I see the following things being beautifully solved by Fastlane:

1. A expressive and extensible DSL.
2. A set of building blocks to take some work from developers.
3. A community of resources and "lanes" whose distribution is taken care of by Bundler through RubyGems.

So the questions that arise on my end are:
- Fastlane's DSL is very Ruby-like, so how much does it make sense to try to port it to Swift vs treating every lane as a CLI with swift-argument-parser as a tool to parse arguments and flags? In the case of the latter, wouldn't Fastlane be more like an automation build tool with SPM at the core?
- What blocks are missing in Swift that would make it difficult for someone to write those workflows in Swift? For example, is "scan" something that needs to be ported? Or should this be taken as an opportunity to revisit the existing layers and peel any unnecessary abstraction? The outcome of this could be a marketplace of recommended Swift packages to automate your workflows.
- What is a "lane" in this new world? Should lanes be executables, and Fastlane just a tool to install and make them discoverable? But in that world, won't Fastlane be taking the responsibilities of a tool like Mise?

The more I think about those questions, the more I think Fastlane's role has the opportunity to evolve into an automation build tool, potentially built on something like [Dagger](https://dagger.io) that blurs automation and CI in a very unique way. Fastlane would add the Swift flavor on top, and sprinkle community of resources and packages on top.

**Comment by mollyIV** [comment]
Closing in favor of the [Fastlane in Swift proposal](https://github.com/fastlane/fastlane/discussions/29570) 🙏
